lib/executor/shared_iterations.go

```diff
@@ -232,15 +234,26 @@ func (si SharedIterations) Run(parentCtx context.Context, out chan<- stats.Sampl
 	regDurationDone := regDurationCtx.Done()
 	runIteration := getIterationRunner(si.executionState, si.logger)
 
+	maxDurationCtx = lib.WithScenarioState(maxDurationCtx, &lib.ScenarioState{
```

Comment:
I don't like how we attach this only to `maxDurationCtx`. Even if this works now, I think this can become a subtle source of bugs in the future, see https://play.golang.org/p/xAEuY_FOOC3

In essence, right now this is the diagram of the contexts currently looks:
```
parentCtx -> maxDurationCtx -> regDurationCtx
                           \-> maxDurationCtxWithScenarioState -> vuCtx
```

This is not necessarily a problem, even before the `vuCtx` was branched as a "sibling" to the `regDurationCtx`, but if we can move the scenario state attachment before the `maxDurationCtx` and `regDurationCtx` it seems to me that it will much safer. Something like this:
```
parentCtx -> parentCtxWithScenarioState -> maxDurationCtx -> regDurationCtx
                                                         \-> vuCtx
```

Reply:
Hhmm would something like [d5ffc781](https://github.com/imiric/k6/commit/d5ffc78183602b87c679aad0854ab2c454efa853) be an improvement?

The problem with attaching it to `parentCtx` is that we need to pass the `progressFn` function to `ScenarioState`. Since this is a lambda created after `getDurationContexts()` is called, I had to split the initialization of `ScenarioState` like this, which is not nice.

A better approach might be to expand the `lib.Executor` interface and refactor all executors to have a standalone `GetProgress()` method (renaming the current one to `GetProgressBar()`), and then attach `ScenarioState` before `Run()` is called:
https://github.com/k6io/k6/blob/f9e0ab6550c21477dbdfcb6cd49be8d755a47f11/core/local/local.go#L329

But that would involve a much larger refactor to avoid what is a hypothetical source of bugs, so I don't want to start that here.

Reply:
Yeah, https://github.com/imiric/k6/commit/d5ffc78183602b87c679aad0854ab2c454efa853 has other issues that might also be a similarly subtle source of future bugs :sweat_smile: And I don't really have a better idea how to not have them without a big executor refactor like the one you propose ... :disappointed:  We could probably have some of this and `getDurationContexts()` wrapped in a new helper function, but the progress will still have to be "dangling" and attached separately.

In any case, we can ignore this for now, but I'll add a new issue to maybe refactor this in the future.

Reply:
Created https://github.com/k6io/k6/issues/2068, resolving this thread so it doesn't clutter the review.